lib/ansible/modules/net_tools/basics/uri.py

```diff
@@ -479,6 +497,57 @@ def form_urlencoded(body):
     return body
 
 
+def form_multipart(body, boundary):
+    ''' Convert data into multipart/form-data '''
+    if isinstance(body, string_types):
+        return body
+
+    if isinstance(body, Mapping):
+        result = []
+        # Convert mapping into multipart chunks
+        for field, value in list(body.items()):
+            # result.extend(('--%s' % boundary))
+            if isinstance(value, string_types):
+                result.extend((
+                    '--%s' % boundary,
+                    'Content-Disposition: form-data; name="%s"' % field,
+                    '',
+                    str(value)
+                ))
+            if isinstance(value, Mapping):
+                # filename required
+                if 'filename' not in value:
+                    raise LookupError('required key "filename" missing from "%s" element' % field)
+                else:
+                    filename = os.path.basename(value['filename'])
+                # set or guess mimetype
+                if 'mimetype' in value:
+                    mimetype = value['mimetype']
+                else:
+                    mimetype = mimetypes.guess_type(filename)[0] or 'application/octet-stream'
+                # get the content or read the file from disk
+                if 'content' in value:
+                    data = value['content']
+                else:
+                    if os.path.exists(value['filename']):
+                        with open(value['filename'], 'rb') as data_file:
+                            data = data_file.read()
+                    else:
+                        raise IOError('unable to read file "%s"' % value['filename'])
+                result.extend((
+                    '--%s' % boundary,
+                    'Content-Disposition: form-data; name="%s"; filename="%s"' % (field, filename),
+                    'Content-Type: %s' % mimetype,
+                    '',
+                    to_text(data)
```

Comment:
If data is binary and not textual, then there is a good chance that passing this through to_text will mangle the binary data.
I just tested this with a wav file, and it seems to mangle it - the textual headers are intact, but the audio content is garbled.
I'll try to think of an alternative for this.

Reply:
I hacked this up using `to_bytes` generously. I dropped the `result` list and instead appended bytes to a `b_result` bytes object. Thus, I could merely append the data to the `b_result` without any additional processing.
My hack was very ugly. I can share it if you like, but it is not clean enough to be merged :P